### PR TITLE
docs: add initial threat model diagrams for multipool autoswap and treasury

### PR DESCRIPTION
**Pull request by jessysaurusrex**
Introducing the initial threat model diagrams for the MultiPool AutoSwap and Treasury.

### CODE REVIEW

**Comment by katelynsills** [comment]
This is awesome! Feel free to add me as a reviewer when you're ready and we can get this merged in!

**Comment by jessysaurusrex** [comment]
@katelynsills Of course! 

This draft set is ready to go -- the main goal is to make them available for others to engage with. Generally, I expect this folder to grow incrementally as more diagrams are ready. It's likely that the the entire effort will be be as "done" as can be considered the accompanying written documentation around risk/mitigation/remediation is merged.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            interface "decrementBy" as usi2
+        }
+
+        interface "reallocate" as reall
+
+    }
+    package "publicFacet" as pf0 {
+        interface "addPool()" as ap
+        interface "otherMethods..." as om
+        interface "offer()" as offer
```
Comment (katelynsills) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml]: I think `offer` should be part of Zoe instead. Is there a reason to put it on the `publicFacet`?

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+pu -> mpas: ""makeAddLiquidityInvitation""
+return ""invitation""
+end
+
+
+group create proposal
+pu -> pu: Create ""proposal"" to add 1 BUCK, and 1 SML\nat ratio of ""poolAllocation"" for X ""liquidity""
+end
+group withdraw funds
+pu -> pu: withdraw funds for ""BUCK"" into ""payments"" map
```
Comment (katelynsills) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I think we should call this an object or record because maps in JavaScript are a different data structure, so it might be confusing. 

```suggestion
pu -> pu: withdraw funds for ""BUCK"" into ""payments"" record
```

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+return result of failure or success
+end
+group recovering funds
+pu -> usA: ""getPayout("in")""
+return ""payoutIn""
+pu -> pu: redeposit ""payoutIn"" into purse
+pu -> usA: ""getPayout("out")""
+return ""payoutOut""
+pu -> pu: redeposit ""payoutOut"" into purse
+end
```
Comment (katelynsills) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: When adding liquidity, the user's offer will have the keywords "Central", "Secondary", and "Liquidity". I think these are the keywords from the swapping offers.

Comment (katelynsills) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: Also a nitpick, but rather than "redeposit" I think we should say "deposit" because there might have been no previous deposit, and never any previous deposit of the same payment.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+psA --> zoe: ""poolSeat"" ref
+zoe -> psA: add/remove funds
+zoe -> usA: add/remove funds
+zoe -> zoe: ""reallocate(seat, poolSeat)""
+
+usA --> pu: ""userSeat"" ref
+pu -> usA: ""getOfferResult()""
+return result of failure or success
+end
+group recovering funds
```
Comment (katelynsills) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml:86]: Same comment here about "In" and "Out" keywords

File: docs/threat_models/treasury/treasury.puml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+actor "TreasuryCreator aka Governance" as tc
+
+actor Borrower
+
+node AdjustBalancesInvitation
+node "CloseVaultInvitation"
+node LiquidationPromise
+node LiquidationSeat {
+    circle getPayoff
+    circle getPayoffs
```
Comment (katelynsills) [inline docs/threat_models/treasury/treasury.puml]: Is there a reason to include both of these?

File: docs/threat_models/treasury/treasury.puml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        circle makeLoanInvitation
+        circle getRunIssuer
+        circle getAMM
+        getAMM -d-> AMM
+    }
+    node "creatorFacet" {
+        circle makeAddTypeInvitation
+    }
+}
+node "Vat" {
```
Comment (katelynsills) [inline docs/threat_models/treasury/treasury.puml:21]: `Vat` wrapping around Vault but not Treasury needs to be changed. All of the vaults are in the same vat as the treasury

**Review by katelynsills — APPROVED** [review]
This looks great! Some small comments

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml (hunk 1 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml]: Technically, yes -- but in order to simplify the diagram and not bring the entirety of Zoe into this context, I opted to represent calling Zoe through the `publicFacet`. We should (and will) diagram all Zoe APIs in a separate diagram pending ongoing work, but this isn't the right place to do that just yet.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 3 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: The details of the offer record are excluded here in order to simplify the diagram. In and Out should encompass the resources from the completed offer result.  

Initially, I was thinking "redeposit" as I wanted to capture the possibility that this may actually have been initial funds. Any ideas on how to capture that in addition to "deposit"?  Failing to deposit these funds would result in their loss, and that would need to be captured somehow.

File: docs/threat_models/treasury/treasury.puml (hunk 5 above)
Comment (jessysaurusrex) [reply docs/threat_models/treasury/treasury.puml]: Both of these methods appear in the Treasury contract, and should be represented independently.  (If they aren't in use, it would be worth considering eliminating one for the sake of simplicity.)

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml (hunk 1 above)
Comment (katelynsills) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml]: Ok, it makes sense to me to leave out Zoe as an object in the diagram for simplification's sake. But, the diagram is going to be misleading in important ways if `offer` is on the public facet of the contract.  The viewer would think that contract code handles offer payments, which could violate offer safety and payout liveness, thus undermining Zoe entirely.  

So, this is something we need to change before this gets merged in. But, I think there's an easy fix, which is an asterisk + comment on the diagram saying that `offer` is actually part of Zoe, and no contract code has access to the user's payments.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 3 above)
Comment (katelynsills) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I think you should use the term `deposit`. `Redeposit` isn't going to be accurate in most cases, especially since the contract can give you more than what you gave. Plus it's shorter :)

About "In" and "Out", I think there's a misunderstanding. What I'm advocating for shouldn't complicate the diagram much. I'm saying replace the word "Out" with "Liquidity", replace the word "In" with "Central", and add a line for an additional payment for "Secondary". Is there any reason not to do that? I think there's no need to generalize the payments since we know what they are, and it's pretty simple to replace the words.

File: docs/threat_models/treasury/treasury.puml (hunk 5 above)
Comment (katelynsills) [reply docs/threat_models/treasury/treasury.puml]: I just searched and I couldn't find anything that looked like `getPayoff` or `getPayoffs`. Where in the contract are you thinking of?

**Comment by katelynsills** [comment]
I've already approved, but adding @Chris-Hibbert in case there is further discussion/questions. Thanks @jessysaurusrex and @Chris-Hibbert!

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    offer --> conf1: execute method
+
+    gps1 -d-> pa1
+    gps2 -d-> pa2
+    conf1 --> pa1i2
+    conf1 --> pa2i2
+    conf1 --> usi1
+    conf1 --> usi2
+    conf1 --> reall
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml]: Is there a reason `conf1` only points to some of the `incrementBy` nodes, though it points to all of the `decrementBy` nodes?

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        [Secondary Token A] --> ct1
+        interface "getPoolSeat()" as gps1
+    }
+
+    node "pool n" as pooln {
+    }
+
+    ap -d-> pooln
+
+    offer --> conf1: execute method
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml:67]: What is this part of the diagram (`offer()`, `validateTerms()`, `executeMethod`) trying to convey? 

Everything else appears to be either the ways a client of the APIs can interact with methods on the AMM pools or how zoe/zcf manage the balances of the pools.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 9]
```diff
@@ -0,0 +1,182 @@
+@startuml
+'https://plantuml.com/sequence-diagram
+header Agoric Operating Systems
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]:
```suggestion
header Agoric Systems Operating Company.
```

(and in the other files

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+actor "Pool User" as pu
+control "Zoe" as zoe
+database "MPAS Contract" as mpas
+
+== MPAS Creation ==
+masc -> mpas: Create instance
+return ""creatorFacet""
+== Pool Addition SML:BCK==
+autonumber 10
+pa -> mpas: ""addPool()""
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml:22]: In order to get from creating an instance to adding a pool, the creator has to publish the instance of the contract somewhere. Otherwise, even though the addPool() method is "public", no one would be able to find it.  It is completely possible in this model to create an AMM, and add pools, and only tell friends about it.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+pu -> pu: Create ""proposal"" to add 1 BUCK, and 1 SML\nat ratio of ""poolAllocation"" for X ""liquidity""
+end
+group withdraw funds
+pu -> pu: withdraw funds for ""BUCK"" into ""payments"" record
+pu -> pu: withdraw funds for ""SML"" into ""payments"" record
+pu -> zoe: ""offer(invitation, proposal, payments)""
+zoe -> zoe: validate offer
+
+zoe -> usA: create ""userSeat""
+usA --> zoe: ""userSeat"" ref
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: There's little need to call out "ref"s in javascript. JS doesn't have a distinction between pointers and objects in the language, so it's easier to just talk about the objects. All references to objects are the same kind of thing.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+zoe -> pool: ""getPoolSeat()""
+pool -> psA: create ""poolSeat"" ref
+psA --> zoe: ""poolSeat"" ref
+
+zoe -> psA: ""addLiquidity""
+zoe -> psA: add/remove funds
+zoe -> usA: add/remove funds
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I would draw a clearer distinction between what `poolSeat.incrementBy()` and `.decrementBy()` are doing in the contract and funds movement. The `reallocate()` call (line 121) actually moves assets between seats, while `incrementBy()` and `.decrementBy()` are moving amounts between staging descriptions. 

No, that's not even right. Zoe holds all the funds in a single purse per brand, and only has allocations for each seat. Zoe is careful to keep the accounts balanced as it moves amounts between accounts, and contracts can't add to balances unless the changes satisfy rights conservation, so the contracts can neither create nor destroy assets, they can only move amounts between accounts maintained by Zoe. When a seat exits, the owner can withdraw the amount the associated balance describes.

Bottom line: we shouldn't talk about moving "funds" when we're talking about actions within a contract.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+zoe -> zoe: ""reallocate(seat, poolSeat)""
+
+usA --> pu: ""userSeat"" ref
+pu -> usA: ""getOfferResult()""
+return result of failure or success
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I think these would be better shown as interactions with Zoe.  Zoe gets the results from ZCF/the contract and can return them to the user even if the contract fails.

[actually, in the context of my later comments, if the user seat is moved to be visibly part of zoe, this would be fine.]

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+create control "userSeat" as usA
+zoe -> usA: create ""userSeat""
+usA --> zoe: ""userSeat"" ref
+zoe -> pool: ""getPoolSeat()""
+create control "poolSeat" as psA
+pool -> psA: create ""poolSeat"" ref
+psA --> zoe: ""poolSeat"" ref
+zoe -> psA: add/remove funds
+zoe -> usA: add/remove funds
+zoe -> zoe: ""reallocate(seat, poolSeat)""
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I think showing this as interacting with Zoe is a mistake. There's a sense in which zoe itself creates the user seat, but it would make more sense to show MPAS as interacting with the pools and directing that balances be changed. When the contract figures out how much to move, it does it by asking zcf to increment or decrement balances on staged allocations, and the `reallocate()` call reaches out from zcf to zoe to update the accounts.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+usA --> pu: ""userSeat"" ref
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I would put UserSeat in a combined (vertical) box with Zoe. Each UserSeat is paired with a ZcfSeat; they're two objects with access to the same state. This diagram shows only one of the zcfSeats that interact during a swap. There's a seat (pair) for each pool, and there's a seat (named `seat` in `swap.js`) that represents the assets the user's assets. The pool seats are persistent through the life of the contract, the user seats are new for each swap (created by the invitation as you show.)


I think `box`, `endbox` lets you draw a box.  see "Participants encompass" [here ](https://plantuml.com/sequence-diagram).

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 3 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: Updated redeposit to deposit.

I see what you mean regarding In/Out -  I updated the terms to reflect what was in the original offer.

File: docs/threat_models/treasury/treasury.puml (hunk 5 above)
Comment (jessysaurusrex) [reply docs/threat_models/treasury/treasury.puml]: Oops I got my wires crossed, `s/Payoff/Payout/g`.  These both appear to be callable in the contract.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml (hunk 7 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml]: Good catch, I've updated them to point correctly.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml (hunk 8 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml:67]: In regards to Zoe, I wanted to show that it referenced the contract and in invoked the contract functionality, however I may have it backwards.  What is the best way to represent `offer()` does stuff with the respective seats here?

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 10 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml:22]: Is it possible to enumerate or directly reference this contract somehow through reflection or other means? Or does it need to be explicitly published?

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 14 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: This makes much more sense, I've updated this so that the contract stages the assets, with Zoe `reallocating`.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 15 above)
Comment (jessysaurusrex) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]: I moved the `userSeat` to be closer to Zoe to hopefully capture this.

File: docs/threat_models/gen.sh [hunk 16]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
```
Comment (Chris-Hibbert) [inline docs/threat_models/gen.sh:1]: Oh, useful!  I'm going to have to crib from this for the governance diagrams.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml (hunk 10 above)
Comment (Chris-Hibbert) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml:22]: It needs to be explicitly published. We have a Board on which references to such things  can be shared, but it takes an explicit action to make them visible. 

It's a crucial part of the security model. We don't want to treat instances of contracts like they're ambient authority, even though we do share them widely for convenience in some circumstances.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+pool2 --> mpas: ""quote""
+mpas --> pu: ""quote""
+end
+
+group making a proposal
+pu -> pu: create proposal for SML:BUCK as ""proposal""
+
+end
+
+group requesting a SwapOut an invitation
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml]:
```suggestion
group requesting an invitation for a SwapOut
```

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+group making an offer
+pu -> pu: withdraw funds from purse of SML into ""payment""
+pu -> zoe: ""offer(invitation, proposal, payment))""
+zoe -> zoe: validate offer
+
+zoe -> usA: create ""userSeat""
+usA --> zoe: ""userSeat""
+zoe -> pool: ""getPoolSeat()""
+create control "poolSeat" as psA
+pool -> psA: create ""poolSeat""
```
Comment (Chris-Hibbert) [inline docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml:76]: There's no "Create poolSeat" call at this point in the transaction. I would actually show the poolSeat as having the same lifetime as the Pool. It's created when the pool comes into existence, and holds the pools assets.

File: docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml (hunk 8 above)
Comment (Chris-Hibbert) [reply docs/threat_models/smart_contracts/multipool_autoswap/mpas_component.puml:67]: Here's a quick sketch. Zoe holds funds in escrow, and has the user seats. each user seat is paired with a ZCF seat in the corresponding ZCF vat.

A ZCF vat runs a single contract (in this case MPAS). MPAS has multiple pools, each of which knows two balances that reflect a claim on escrowed funds in Zoe.

When a seat is created for a transaction (`addLiquidity` or `adjustBalances` are the two examples here) it interacts with the conctract code, reallocations are done, and the user gets something back. 

The publicFacet, creatorFacet, etc. could also be pictured within the ZCF vat, since they're facets of the contract.

Threat Model--multipool components

We can jump on a call if you want to wave my hands and point at the parts of the diagram.
